File: gherkin/java/src/main/java/gherkin/TokenFormatter.java
package gherkin;

public class TokenFormatter {
    private static final StringUtils.ToString<GherkinLineSpan> SPAN_TO_STRING = new StringUtils.ToString<GherkinLineSpan>() {
        @Override
        public String toString(GherkinLineSpan o) {
            return o.column + ":" + o.text;
        }
    };

    public String formatToken(Token token) {
        if (token.isEOF())
            return "EOF";

        return String.format("(%s:%s)%s:%s/%s/%s",
                toString(token.location.getLine()),
                toString(token.location.getColumn()),
                toString(token.matchedType),
                toString(token.matchedKeyword),
                toString(token.matchedText),
                toString(token.mathcedItems == null ? "" : StringUtils.join(SPAN_TO_STRING, ",", token.mathcedItems))
        );
    }

    private String toString(Object o) {
        return o == null ? "" : o.toString();
    }
}

File: gherkin/php/src/TokenFormatter.php
<?php

declare(strict_types=1);

namespace Cucumber\Gherkin;

final class TokenFormatter
{
    public function formatToken(Token $token): string
    {
        if ($token->isEof()) {
            return 'EOF';
        }

        return sprintf(
            "(%s:%s)%s:%s/%s/%s",
            $token->getLocation()->line,
            $token->getLocation()->column,
            $token->match?->tokenType->name ??  '',
            $token->match?->keyword ? $this->formatKeyword($token) : '',
            $token->match?->text ?? '',
            $token->match === null ? ''
                : join(',', array_map(fn ($linespan) => $linespan->column . ':' . $linespan->text, $token->match->items)),
        );
    }

    private function formatKeyword(Token $token): string
    {
        return sprintf(
            "(%s)%s",
            $token->match?->keywordType->value ?? '',
            $token->match?->keyword ?? '',
        );
    }
}
